File: architecture-skills/design-patterns/examples/adapter/adapter.php
<?php
// Adapter pattern in PHP.
//
// The adapter implements the target interface and delegates to the
// wrapped legacy instance, converting units at the boundary.

interface TemperatureSource
{
    public function name(): string;
    public function celsius(): float;
}

class ModernSensor implements TemperatureSource
{
    public function __construct(
        private string $name,
        private float $celsius,
    ) {}

    public function name(): string
    {
        return $this->name;
    }

    public function celsius(): float
    {
        return $this->celsius;
    }
}

// Adaptee: third-party class we cannot modify.
class LegacyFahrenheitSensor
{
    public function __construct(
        public readonly string $sensorId,
        private float $fahrenheit,
    ) {}

    public function readFahrenheit(): float
    {
        return $this->fahrenheit;
    }
}

class FahrenheitAdapter implements TemperatureSource
{
    public function __construct(private LegacyFahrenheitSensor $legacy) {}

    public function name(): string
    {
        return $this->legacy->sensorId;
    }

    public function celsius(): float
    {
        return ($this->legacy->readFahrenheit() - 32.0) * 5.0 / 9.0;
    }
}

/** @param TemperatureSource[] $sources */
function report(array $sources): void
{
    foreach ($sources as $source) {
        printf("%s: %.1f°C\n", $source->name(), $source->celsius());
    }
}

report([
    new ModernSensor('server-room', 21.5),
    new FahrenheitAdapter(new LegacyFahrenheitSensor('warehouse-legacy', 77.0)),
]);

File: architecture-skills/design-patterns/examples/adapter/adapter.cs
// Adapter pattern in C#.
//
// An object adapter: the adapter implements the target interface and
// delegates to the wrapped legacy instance. System.IO.StreamReader
// (stream -> text) plays the same role in the base class library.

using System;
using System.Collections.Generic;

interface ITemperatureSource
{
    string Name { get; }
    double Celsius { get; }
}

class ModernSensor : ITemperatureSource
{
    public string Name { get; }
    public double Celsius { get; }

    public ModernSensor(string name, double celsius)
    {
        Name = name;
        Celsius = celsius;
    }
}

// Adaptee: third-party class we cannot modify.
class LegacyFahrenheitSensor
{
    public string SensorId { get; }
    private readonly double _fahrenheit;

    public LegacyFahrenheitSensor(string sensorId, double fahrenheit)
    {
        SensorId = sensorId;
        _fahrenheit = fahrenheit;
    }

    public double ReadFahrenheit() => _fahrenheit;
}

class FahrenheitAdapter : ITemperatureSource
{
    private readonly LegacyFahrenheitSensor _legacy;

    public FahrenheitAdapter(LegacyFahrenheitSensor legacy) => _legacy = legacy;

    public string Name => _legacy.SensorId;
    public double Celsius => (_legacy.ReadFahrenheit() - 32.0) * 5.0 / 9.0;
}

class Program
{
    static void Main()
    {
        var sources = new List<ITemperatureSource>
        {
            new ModernSensor("server-room", 21.5),
            new FahrenheitAdapter(new LegacyFahrenheitSensor("warehouse-legacy", 77.0)),
        };
        foreach (var source in sources)
        {
            Console.WriteLine($"{source.Name}: {source.Celsius:F1}°C");
        }
    }
}
